### PR TITLE
core: RFC: logging API

### PR DESCRIPTION
**Pull request by kaspar030**
my take on a logging API. Compare #2503.

- introduces log levels error < warning < info < debug and corresponding log_* with printf semantics
- logging calls will be optimized out if LOG_LEVEL is defined lower than a certain level.
- LOG_LEVEL=LOG_NONE disables all logging
- any module can overwrite built-in logging by providing a "log_module.h" and setting "-DHAVE_LOG_MODULE"
- every logging call is syntax checked
- API allows macros, static inline functions as well as normal functions to do the actual work

This could replace ```debug.h``` as LOG_LEVEL can be redefined on a per-file basis.

In order to use this, add "-DLOGLEVEL=something" to your applications CFLAGS, with "something" being any of the log levels defined in log.h.

This is a discussion PR without any documentation.

### CODE REVIEW

**Comment by kaspar030** [comment]
I've changed the api from ```log_info(stuff);``` to ```log(LOG_INFO, stuff);```. While the define is ugly, this makes the API even simpler.

**Comment by kaspar030** [comment]
We could use a mask instead of an integer level, in order to be able to select specific log entries. That would enable subsystems to define their own togglable log streams.

**Comment by authmillenon** [comment]
I don't really like the addition of new include paths nor that the public `.h` file is in the module's code directory.

File: core/include/log.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    LOG_NONE,
+    LOG_ERROR,
+    LOG_WARNING,
+    LOG_INFO,
+    LOG_DEBUG,
+    LOG_ALL
+};
+
+#ifndef LOG_LEVEL
+#define LOG_LEVEL LOG_ALL
```
Comment (authmillenon) [inline core/include/log.h]: Log info is a much nicer default level :).

File: core/include/log.h [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+ * @{
+ *
+ * @file        log.h
+ * @brief       System logging header
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ */
+
+#ifndef __LOG_H
+#define __LOG_H
```
Comment (authmillenon) [inline core/include/log.h]: No l leading underscores in guards anymore.  See https://github.com/RIOT-OS/RIOT/pull/2412#discussion_r24337496

**Comment by authmillenon** [comment]
In general I like the idea of log levels better, but the overriding mechanism of @LudwigOrtmann's PR is cleaner.

**Comment by kaspar030** [comment]
@authmillenon The header could move into "include/".

While I agree @LudwigOrtmann's overriding mechanism is a little cleaner, it makes the use of macros or static inlines within logging modules impossible.

File: sys/log/log_printfnoformat/log_module.h [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#ifndef __LOG_FORMAT_H
+#define __LOG_FORMAT_H 
+
+#include <stdio.h>
+
+static void _log(unsigned level, const char *format, ...) {
```
Comment (authmillenon) [inline sys/log/log_printfnoformat/log_module.h]: `inline` is missing

**Comment by authmillenon** [comment]
> @authmillenon The header could move into "include/".

Wouldn't that lead to a lot of filenamw collisions? 

> While I agree @LudwigOrtmann's overriding mechanism is a little cleaner, it makes the use of macros or static inlines within logging modules impossible

What is speaking against his approach but making `log_printfnoformat` a pseudo-module with your header definitions?

**Comment by kaspar030** [comment]
>> @authmillenon The header could move into "include/".

> Wouldn't that lead to a lot of filenamw collisions?

Why? the include path is only added if that log module is to be used.

**Comment by kaspar030** [comment]
> What is speaking against his approach but making log_printfnoformat a pseudo-module with your header definitions?

Not sure what you mean by "his approach". @LudwigOrtmann doesn't include a logging module's header, but uses either builtin macros or declares a function that a log module can implement. That way a log module cannot offer a static inline log function or a plain macro.

**Comment by authmillenon** [comment]
Why is the branch in the upstream repository btw?

**Comment by OlegHahm** [comment]
valid question ;)

**Comment by kaspar030** [comment]
Sorry for that. Should I delete it right away?

**Comment by authmillenon** [comment]
I guess moving it to your own repository and opening a new pull request is also necessary. And please do

```bash
git remote set-url \
    --push "$(git remote -v | grep "/RIOT-OS/RIOT\(.git\)* (push)" | awk '{print $1}')" \
    git://github.com/RIOT-OS/RIOT.git
```

**Comment by authmillenon** [comment]
This way this can't accidentally happen again ;-)

**Comment by kaspar030** [comment]
well, the problem was a fresh checkout directly from  git://github.com/RIOT-OS/RIOT.git, so fixing my .git/config won't help against these mistakes... usually I check out my own clone and add upstream as remote. That push was a 3a.m. mistake...

**Comment by kaspar030** [comment]
Ok, deleted, wait for a version out of my fork.

**Pull request closed without merging**